src/lib.rs

```diff
@@ -0,0 +1,386 @@
+        if !init() {
+            whatever!("unable to initialize magewell api");
+        }
+
+        let handle = {
+            let _lock = DEVICE_LIST_MUTEX
+                .lock()
+                .whatever_context("unable to lock device list")?;
+
+            unsafe {
+                if sys::MWRefreshDevice() != sys::_MW_RESULT__MW_SUCCEEDED {
+                    whatever!("unable to refresh device list");
+                }
+            }
+
+            let handle = unsafe { sys::MWOpenChannel(board_index as _, channel_index as _) };
+            if handle.is_null() {
+                whatever!("unable to open channel");
+            }
+            handle
+        };
+
+        let info: ChannelInfo = {
+            let mut info = MaybeUninit::uninit();
+            unsafe {
+                if sys::MWGetChannelInfo(handle, info.as_mut_ptr()) != sys::_MW_RESULT__MW_SUCCEEDED
+                {
+                    sys::MWCloseChannel(handle);
+                    whatever!("unable to get channel info");
+                }
+                info.assume_init().into()
+            }
+        };
+
+        Ok(if info.family_name().to_str() == Ok("Eco Capture") {
+            let event_fd = EventFd::new().whatever_context("unable to create eventfd")?;
+            Channel::Eco(EcoChannel {
+                handle,
+                info,
+                event_fd,
+                video_capture_frame: None,
+            })
+        } else {
+            Channel::Pro(ProChannel { handle, info })
+        })
+    }
+}
+
+pub trait UniversalCaptureFamily {
+    fn handle(&self) -> *mut c_void;
+
+    fn info(&self) -> &ChannelInfo;
+
+    fn get_video_signal_status(&self) -> Result<VideoSignalStatus> {
+        let mut status = MaybeUninit::uninit();
+        unsafe {
+            if sys::MWGetVideoSignalStatus(self.handle(), status.as_mut_ptr())
+                != sys::_MW_RESULT__MW_SUCCEEDED
+            {
+                whatever!("unable to get video signal status");
+            }
+            Ok(status.assume_init().into())
+        }
+    }
+}
+
+pub trait ProEcoCaptureFamily: UniversalCaptureFamily {
+    fn get_device_time(&self) -> Result<Duration> {
+        let mut time: c_longlong = 0;
+        unsafe {
+            if sys::MWGetDeviceTime(self.handle(), &mut time as *mut _)
+                != sys::_MW_RESULT__MW_SUCCEEDED
+            {
+                whatever!("unable to get device time");
+            }
+            Ok(Duration::from_nanos(100 * time as u64))
+        }
+    }
+}
+
+impl UniversalCaptureFamily for Channel {
+    fn handle(&self) -> *mut c_void {
+        match self {
+            Channel::Eco(ch) => ch.handle(),
+            Channel::Pro(ch) => ch.handle(),
+        }
+    }
+
+    fn info(&self) -> &ChannelInfo {
+        match self {
+            Channel::Eco(ch) => ch.info(),
+            Channel::Pro(ch) => ch.info(),
+        }
+    }
+}
+
+pub struct EcoChannel {
+    handle: *mut c_void,
```

Comment:
I think I'd create a `ChannelHandle` wrapper, with its `Drop` impl replacing the three places where channels are closed now: an error path in `Channel::open`, `<EcoChannel as Drop>::drop`, `<ProChannel as Drop>::drop`. The other `Drop` impls wouldn't be needed at all then, and there'd be no worries about accidentally introducing a leak if there's more error paths added to `Channel::open` or more channel types.

Reply:
Good idea. 👍